Adding a file_browser.html example

telamonian opened the pull request:
Adds an example of regular-table being used to render a filebrowser, complete with a tree column on the left. Still needs a bunch of work to be functional, and then a whole bunch of styling work beyond that. I'd also like to figure out how to add some "sort on header click" behavior.

I want to explore adding multiselect and drag-n-drop behavior to this example filebrowser, but those are complex and I may end up saving them for their own individual examples.

telamonian commented:
This PR currently relies on a filthy hack that I inserted into `tree_header`:

https://github.com/jpmorganchase/regular-table/blob/0bf670c0881d53914bf975cb19eb0a974d5c20c4/src/js/tree_row_header.js#L42-L44

There's probably a better/more generic way to arbitrarily mark rows as "not a leaf". I'll see if anything occurs to me after I've rebased this PR on current master (including #17)

telamonian commented:
Basic functionality achieved!

![image](https://user-images.githubusercontent.com/2263641/82846465-f3d0b800-9eb6-11ea-8c1b-4bbd55086d3c.png)

For a minute I thought I was going to have to implement `expand` using an indexed priority queue or some crazy thing. Then I realized I was waaaay overthinking it, and came up with something nicely simple.

Still needs a ton of styling work

telamonian commented:
Basic functionality achieved!

![image](https://user-images.githubusercontent.com/2263641/82846465-f3d0b800-9eb6-11ea-8c1b-4bbd55086d3c.png)

For a minute I thought I was going to have to implement `expand` using an indexed priority queue or some crazy thing. Then I realized I was waaaay overthinking it, and came up with something nicely simple.

Still needs a ton of styling work

telamonian commented:
I fixed an issue with the icon font ligatures, so now the example looks a bit nicer:

![image](https://user-images.githubusercontent.com/2263641/82859772-9e5fcf80-9ee5-11ea-889c-a6a58b17e4e4.png)

There's a couple more things I'd like to go into this example (file/dir icons, sort-on-header-click), but I'd also like to avoid having to do another rebase. So I think this PR is good to go as-is.

The only bone of contention may be the tweaks I made to how "leafness" is determined for certain rows. @texodus Do you have any more clever suggestions for how I could explicitly specify which rows are not leafs?

texodus commented on examples/file_browser.html:
```diff
@@ -0,0 +1,119 @@
+
+        const NUM_ROWS = 100;
+        const NUM_DIR = 10;
+        const NUM_FILE = NUM_ROWS - NUM_DIR;
```
Unused `NUM_FILE`

telamonian replied:
thanks for the spot! hangover from earlier code

telamonian replied:
done!

texodus commented on examples/file_browser.html:
```diff
@@ -0,0 +1,119 @@
+            };
+        }
+
+        // for tree support
```
`collapse` and `expand` are perspective API holdovers that will be removed.  
* For the time being, they should be refactored to `__collapse` and `__expand` respectively, and please also apply this fix to `perspective.html` to map these back to `perspective.View()`.
* Eventually, use `.addEventListener("click", =>` on the `regular-table` and check for button contact in the example.
* In the far future, remove the +/- rendering entirely and implement this as a custom cell with templating.

Also, can we simplify this by replacing inlining and removing the ambiguously-named `popContents()` and `insertContents()`?

telamonian replied:
Done! I added `__collapse`, `__expand`, and `__set_depth` to `perspective.html`, and added corresponding code in `RegularVirtualTableViewModel.draw` in `scroll_panel.js`.

I ran into an issue with the perspective example where all of the row indices were rendering as `NaN`. This turned out to be because all of the columns types listed in `__schema` were `integer`, and the tree row header tries to format the row header text based on that type:

https://github.com/jpmorganchase/regular-table/blob/50a7ea908d91cb969961b2bdd00262ef5c8f3229/src/js/tree_row_header.js#L43

All of this formatting stuff is going to get thrown out or significantly refactored, right? For right now, I was able to "fix" this by setting `__schema` to `table_schema` in `perspective.html`. @texodus Does this seem like a reasonable temporary fix to you?

I also took a look at the original code taken from perspective. There, this formatting is done on the basis of `table_schema`, and not the view's `schema`.

texodus commented on src/js/tree_row_header.js:
```diff
@@ -38,6 +38,11 @@ function _tree_header_levels(path, is_open, is_leaf) {
     const type = types[path.length - 1];
     const name = path.length === 0 ? "TOTAL" : path[path.length - 1];
+
+    if (path.length && path[path.length - 1].endsWith("/")) {
```
This doesn't belong here.  I'd like to hear a proposal on how to handle tree rendering as part of the data model API, rather than bash this implementation into submission, especially as there are tree forms the current renderer does not support.

telamonian replied:
I agree, although this was nicely quick and easy. I'll think about this. A couple of idea off the top of my head:

1. row, col, cell, etc metadata that can be explicitly provided by the data model
2. make the data model itself hierarchical

I played around with 2. a bit while I was originally implementing the tree for file_browser.html. It's probably not a great idea, since it completely screws up the concept/utility of row index, but maybe there's a way to do an appropriate nested data structure here?

texodus commented on examples/file_browser.html:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        const table = document.getElementsByTagName('regular-table')[0];
+        table.setDataModel(file_browser_model).then(() => {
```
Please use `async`/`await` style in examples, unless in the odd case, a `Promise` clearly makes it more readable.  Also, I think this skips 1 frame on initial render?

telamonian replied:
I was not able to `await` here at the top level of the html-embedded script (runtime error), which is why I switched to promise syntax. Is there a way to make `await` work in this context?

texodus commented on examples/file_browser.html:
```diff
@@ -0,0 +1,119 @@
+
+        const table = document.getElementsByTagName('regular-table')[0];
+        table.setDataModel(file_browser_model).then(() => {
+            table._view_cache.config = {
```
Use `__config` to indicate this is legacy functionality bound for removal, ala `perspective.html` example.  This is not really how `row_pivot` parameter works, and is necessitates the `is_leaf` tweak below.

telamonian replied:
I was trying to figure out a better way to pass in the name of the `row_indices` column. My proposal is that the `col`/`row_indices` arrays be replaced by Index objects (a la [Pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Index.html)) of the form:

```
{
  name: string;
  values: string[][];
}
```

Some simpler (but problematic) alternative proposals:
- expect an optional `.name` property on the existing `row_indices` array
- if `row_indices` is present, expect to find its name at `column_indices[0]` (ie `column_indices.length` should be `data.length + 1` when `row_indices` is present)

telamonian replied:
I am now setting `row_pivots` via `__config`. However, right now `row_pivots` has no visible effect on the regular-table rendered by the `file_browser.html` example. The whole reason why I wanted to set `row_pivots` in the first place was to set the header text on the `row_indices` column, as implied here in the source:

https://github.com/jpmorganchase/regular-table/blob/ec0c5742f359170cd027fa901eb7a31b13003247/src/js/table.js#L84-L95

However, since `column_name` gets passed as the second arg of `header.draw`, that means it gets set as the column's `alias`, not as the actual displayed column name.

@texodus What's the distinction here between alias and actual column name? Is this something we can simplify for regular-table?

texodus commented on src/js/tbody.js:
```diff
@@ -52,7 +52,7 @@ export class RegularBodyViewModel extends ViewModel {
             metadata.row_path = null;
         } else if (formatter) {
-            formatter.format(td, val, type, val.length === depth, is_open);
+            formatter.format(td, val, type, val.length >= depth, is_open);
```
Breaks perspective integration, which is not currently enabled in the test suite.  Formatting is another feature destined for the trash heap regardless, as this is better handled within the data model.

telamonian replied:
Can you please help me understand the issue here a little better? The `perspective.html` example in the latest version of this PR does seem to correctly render the tree in its first column, so it would be helpful to know exactly what this line breaks.

texodus commented on examples/file_browser.html:
```diff
@@ -0,0 +1,119 @@
+        const formatter = new Intl.DateTimeFormat("en-us");
+
+        const _contents_cache = new Map();
+        function getContents(path, expand) {
```
This is clever - but it would benefit from being cleaned up, and maybe structurally separated to a logically distinct `generateFakeFilesystem()` method so its more clear how this examples uses the `regular-table` API.

telamonian replied:
I'm fine with the name change, but I don't quite follow the rest of your comment here. Can you please add some more details and/or concrete examples of what you'd like to see changed?

telamonian commented:
@texodus Response to review:

- [ x ] addressed all of the review comments and/or asked for follow up
- [ x ] added tests
- [ x ] added more specific tests for the tree column
- [ x ] create a file_browser bl.ocks and added it to the README

I'll leave the commit squash for last, after the review is settled.

@texodus Can you please take another look at this PR now? I'll need some further input in order to address the remaining issues from the review

telamonian commented:
The latest commit (c1f0cc0) added filetype icons to the file browser example:

![image](https://user-images.githubusercontent.com/2263641/83082551-bc4a4300-a051-11ea-95cd-2b00964218ab.png)

Much more welcoming. The `"regular-table-after-update"` listener that paints the icons could use some cleanup.

tgeorgeux commented:
@telamonian is there an easy way to install/test this? I was trying to show this to the UCI team and I realized I have no idea how to get it running. For JupyterLab I have a bash script that I wrote 2 years ago to install and run PRs. 😂

texodus commented:
Thanks! This has been rebased and merged in #25
